### ISSUE TITLE
Error while compiling thrust/complex.h with clang version 9.0.0

### ISSUE THREAD

**Issue by hbrunie** [thrust/thrust#985]
Hi everybody,

I tried to compile HelloWord code including thrust/complex.h with clang.

`clang++ --version`
`clang version 9.0.0`
(<https://github.com/llvm-mirror/clang> 771c454fcb9a98b83da46bd29eb4cd090092b5da) 
(<https://github.com/llvm-mirror/llvm> c5a9fd42692e6e05f647904f4e5ed8851139d27d)
`Target: x86_64-unknown-linux-gnu`
`Thread model: posix`

Trying to compile this simple.cu:
`#include <thrust/complex.h>`
`int main(int argc, char* argv[]) {`
 `   return 0;`
`}`

with this compile command:
`clang++ -I$CUDA_ROOT/samples/common/inc simple.cu -o simple.exe --cuda-gpu-arch=sm_70 -L$CUDA_ROOT/lib64 -lcudart`

I obtained this compiler error:
`In file included from simple.cu:1:
In file included from /[...]/cuda/9.2/include/thrust/complex.h:973:
/[...]/cuda/9.2/include/thrust/detail/complex/complex.inl:193:25: error: __host__ function 'operator=' cannot overload __host__ __device__ function 'operator='
complex<T>& complex<T>::operator=(const complex<T>& z)
                        ^
/[...]/cuda/9.2/include/thrust/complex.h:180:12: note: previous declaration is here
  complex& operator=(const complex<T>& z);
           ^
1 error generated when compiling for sm_70.`

This error seems to be related to the difference of dialects between what nvcc and clang can understand: [overloading-based-on-host-and-device-attributes ](http://llvm.org/docs/CompileCudaWithLLVM.html#overloading-based-on-host-and-device-attributes) or [using-a-different-class-on-host-device ](http://llvm.org/docs/CompileCudaWithLLVM.html#using-a-different-class-on-host-device).

Some more information:

`uname -a
Linux cgpu04 4.4.165-81-default #1 SMP Mon Dec 3 07:47:23 UTC 2018 (0bfc9ed) x86_64 x86_64 x86_64 GNU/Linux`

`srun nvidia-smi 
Wed May  8 13:55:33 2019       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 418.40.04    Driver Version: 418.40.04    CUDA Version: 10.1     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  Tesla V100-SXM2...  Off  | 00000000:1A:00.0 Off |                    0 |
| N/A   27C    P0    37W / 300W |      0MiB / 16130MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU       PID   Type   Process name                             Usage      |
|=============================================================================|
|  No running processes found                                                 |
+-----------------------------------------------------------------------------+
`

Would you like more information? What is wrong with my thrust installation?
Thanks,
Hugo

PS: I had same problem with llvm 7.0.1 and CUDA 9.x (x==0 || x==1 || x==2)

**Comment by griwes** [comment]
If I understand correctly, the issue comes from the fact that that _definition_ of `operator=` didn't have `__host__ __device__` on it; that is no longer the case, since 2739a5535c7a972ad482ffb12e63df1fa749b0ea (the removed lines at the end removed a function, but at the same time left the `__host__ __device__` on the next remaining function; it took me a while to notice what's going on there).

In other words, this is fixed since 10.1.

**Comment by hbrunie** [comment]
I guess we just have to update our system! Thank you.
